core/src/sigverify_shreds.rs

```diff
@@ -0,0 +1,395 @@
+use crate::cuda_runtime::PinnedVec;
+use crate::packet::{Packet, Packets};
+use crate::recycler::Recycler;
+use crate::recycler::Reset;
+use crate::sigverify::{batch_size, TxOffset};
+use bincode::deserialize;
+use rayon::iter::IntoParallelIterator;
+use rayon::iter::ParallelIterator;
+use rayon::ThreadPool;
+use solana_ledger::perf_libs;
+use solana_ledger::shred::ShredType;
+use solana_metrics::inc_new_counter_debug;
+use solana_rayon_threadlimit::get_thread_count;
+use solana_sdk::pubkey::Pubkey;
+use solana_sdk::signature::Signature;
+use std::collections::HashMap;
+use std::mem::size_of;
+
+use std::cell::RefCell;
+
+thread_local!(static PAR_THREAD_POOL: RefCell<ThreadPool> = RefCell::new(rayon::ThreadPoolBuilder::new()
+                    .num_threads(get_thread_count())
+                    .thread_name(|ix| format!("sigverify_shreds_{}", ix))
+                    .build()
+                    .unwrap()));
+
+impl Reset for PinnedVec<Pubkey> {
+    fn reset(&mut self) {
+        self.resize(0, Pubkey::default());
+    }
+}
+
+/// Assuming layout is
+/// signature: Signature
+/// signed_msg: {
+///   type: ShredType
+///   slot: u64,
+///   ...
+/// }
+/// Signature is the first thing in the packet, and slot is the first thing in the signed message.
+fn verify_shred_cpu(packet: &Packet, slot_leaders: &HashMap<u64, Pubkey>) -> Option<u8> {
+    let sig_start = 0;
+    let sig_end = size_of::<Signature>();
+    let slot_start = sig_end + size_of::<ShredType>();
+    let slot_end = slot_start + size_of::<u64>();
+    let msg_start = sig_end;
+    let msg_end = packet.meta.size;
+    trace!("slot start and end {} {}", slot_start, slot_end);
+    let slot: u64 = deserialize(&packet.data[slot_start..slot_end]).ok()?;
```

Comment:
I'm not sure. However, there is no need to check bounds checks like [here](https://github.com/solana-labs/solana/blob/8cf7aa48609025d96d34b4d4382c9e72499c6ad5/core/src/sigverify.rs#L101), isn't there?